x-pack/plugins/observability_ai_assistant/server/service/knowledge_base_service/index.ts

```diff
@@ -291,64 +292,202 @@ export class KnowledgeBaseService {
     }
   };
 
+  private async recallFromKnowledgeBase({
+    queries,
+    contexts,
+    namespace,
+    user,
+    modelId,
+  }: {
+    queries: string[];
+    contexts?: string[];
+    namespace: string;
+    user: { name: string };
+    modelId: string;
+  }): Promise<RecalledEntry[]> {
+    const query = {
+      bool: {
+        should: queries.map((text) => ({
+          text_expansion: {
+            'ml.tokens': {
+              model_text: text,
+              model_id: modelId,
+            },
+          } as unknown as QueryDslTextExpansionQuery,
+        })),
+        filter: [
+          ...getAccessQuery({
+            user,
+            namespace,
+          }),
+          ...getCategoryQuery({ contexts }),
+        ],
+      },
+    };
+
+    const response = await this.dependencies.esClient.search<
+      Pick<KnowledgeBaseEntry, 'text' | 'is_correction' | 'labels'>
+    >({
+      index: [this.dependencies.resources.aliases.kb],
+      query,
+      size: 20,
+      _source: {
+        includes: ['text', 'is_correction', 'labels'],
+      },
+    });
+
+    return response.hits.hits.map((hit) => ({
+      ...hit._source!,
+      score: hit._score!,
+      id: hit._id,
+    }));
+  }
+
+  private async recallFromConnectors({
+    queries,
+    asCurrentUser,
+    modelId,
+  }: {
+    queries: string[];
+    asCurrentUser: ElasticsearchClient;
+    modelId: string;
+  }): Promise<RecalledEntry[]> {
+    const ML_INFERENCE_PREFIX = 'ml.inference.';
+
+    const fieldCaps = await asCurrentUser.fieldCaps({
```

Comment:
Alarm bells on the speed of this call?

Reply:
I think it's fine. It does cause issues with a huge number of mapping, and cross-cluster search, but I suspect neither of them are applicable here. I think it's fine for now. In the future we might want to add an opt-out for users (which would be another config option).